presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/file/FileHiveMetastore.java

```diff
@@ -990,11 +994,24 @@ private boolean isValidPartition(Table table, String partitionName)
     }
 
     @Override
-    public synchronized Optional<List<String>> getPartitionNamesByParts(HiveIdentity identity, String databaseName, String tableName, List<String> parts)
+    public Optional<List<String>> getPartitionNamesByFilter(
+            HiveIdentity identity,
+            String databaseName,
+            String tableName,
+            List<String> columnNames,
+            TupleDomain<String> partitionKeysFilter)
     {
-        // todo this should be more efficient by selectively walking the directory tree
-        return getPartitionNames(identity, databaseName, tableName).map(partitionNames -> partitionNames.stream()
-                .filter(partitionName -> partitionMatches(partitionName, parts))
+        if (partitionKeysFilter.isNone()) {
+            return Optional.of(ImmutableList.of());
+        }
+        Optional<List<String>> parts = partitionKeyFilterToStringList(columnNames, partitionKeysFilter, assumeCanonicalPartitionKeys);
```

Comment:
We shouldn't need the `assumeCanonicalPartitionKeys` config or filtering here at all, as filtering by the metastore is best effort. This isn't being pushed down, so we can let the caller do it.

Reply:
isn't `assumeCanonicalPartitionKeys` applicable to some metastores and not the others  -- eg depending whether the call is over string representation or interpreted values?

for example it may turn out to be HMS- (and file-?) metastore specific?